### PR TITLE
Only set Ticket ref if not yet present via import or synchro

### PR DESCRIPTION
**Pull request by Hipska**
As requested by [#1475 Only generate ticket ref when empty](https://sourceforge.net/p/itop/tickets/1475/).

### CODE REVIEW

**Comment by Hipska** [comment]
Hi, this PR has been requested by @v-dumas, but no response from Combodo yet..

**Comment by piRGoif** [comment]
Hello Thomas, 
We are currently working hard on 2 new versions, we will certainly do a review of the PR on friday, so hopefully there should be some news next week :)

**Comment by piRGoif** [comment]
Hello !
Indeed your modification makes perfect sense.
But it introduces a major change as before ref would always depends on object id. This may leads to some uniqueness problem... Could you add a uniqueness constraint on the ref ?

**Comment by Hipska** [comment]
Hi @piRGoif, you are free to do it. Maintainers can do edits.

**Comment by piRGoif** [comment]
>> Could you add a uniqueness constraint on the ref ?
>
> Hi @piRGoif, you are free to do it. 

I ask you to add it :) : this modification as it is now can lead to integrity issues, so it can't be merged without modification.

> Maintainers can do edits.

Didn't understand what you meant ?

**Comment by Hipska** [comment]
I meant this Github feature:

- [x] Allow edits from maintainers.
> If checked, **users with write access to Combodo/iTop can add new commits** to your `feature/ticket_ref_overwrite` branch. You can always change this setting later.

**Comment by piRGoif** [comment]
Hello Thomas,

Oh ok I didn't knew this option, it's cool !-) There is a whole documentation here : https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

The use case that could cause problems is importing tickets in a brand new iTop, for example refs 10 to 15, then creating new tickets (beginning from ref 1) : you'll get problems when creating the 10th ticket. The solution would be to change the base for the auto increment id field, an operation that needs to be performed by the db admin, so not really easy to do...
Having a blocking uniqueness rule would be painful for users, so I'll add a non blocking rule.

**Comment by piRGoif** [comment]
> The use case that could cause problems is importing tickets in a brand new iTop, for example refs 10 to 15, then creating new tickets (beginning from ref 1) : you'll get problems when creating the 10th ticket.

As @bruno-ds pointed out in a private message, this is wrong !

See https://dev.mysql.com/doc/refman/5.7/en/example-auto-increment.html :

> When you insert any other value into an AUTO_INCREMENT column, the column is set to that value and the sequence is reset so that the next automatically generated value follows sequentially from the largest column value.

But hey, we will keep anyway the uniqueness rule non blocking, to be extra careful :)

**Comment by Hipska** [comment]
That would indeed be no problem if that user also imported ticket id. But I guess they would not do that most of the time. So a uniqueness rule would still be useful.

**Comment by Molkobain** [comment]
Yeah, +1 with @Hipska . We said that because most of the time IDs are not exported, so we can't be sure they will be imported with it.

**Comment by piRGoif** [comment]
Yes, you're perfectly right, I forgot to mention this would be the case if id were imported also. But you did understand anyway :)

**Comment by piRGoif** [comment]
I've just pushed the uniqueness rule on your fork, then squash&merge !
Thanks

**Comment by Hipska** [comment]
I was just thinking, even if you are also importing ticket ID, even then the ref would be calculated on the next available ID, the method does not take into account a current ID. Which is also the reason this breaks on MySQL 8.

**Comment by piRGoif** [comment]
Hello,
Yes you're right : in \CMDBSource::GetNextInsertId we do SHOW TABLE STATUS and get the auto_increment value, so the max(id) + 1.

**Comment by piRGoif** [comment]
Oh, by the way, we aim to fix MySQL8 incompatibilities in 2.7.0 O:)

File: datamodels/2.x/itop-tickets/datamodel.itop-tickets.xml [hunk 1]
```diff
@@ -53,6 +53,14 @@
 				   </attributes>
 				 </index>
 				</indexes>
+				<uniqueness_rules>
+					<rule if="ref">
```
Comment (Hipska) [inline datamodels/2.x/itop-tickets/datamodel.itop-tickets.xml]: Still needs to be changed to `id`

Comment (Hipska) [reply datamodels/2.x/itop-tickets/datamodel.itop-tickets.xml]: Nevermind, just saw this is already merged and fixed..

Comment (piRGoif) [reply datamodels/2.x/itop-tickets/datamodel.itop-tickets.xml]: Thanks for your comment Thomas - indeed as you saw Jeffrey already sent the same feedback. Sorry for this silly mistake :/

**Pull request merged**